### ISSUE TITLE
Change `Application` to use `Aff` rather than `Effect`

### ISSUE THREAD

**Issue by JordanMartinez** [Woody88/purescript-wai#5]
When I look at the [Haskell version of wai](https://hackage.haskell.org/package/wai-3.2.2.1/docs/Network-Wai.html), they use `IO` as their monad. The corresponding monad in PureScript is not `Effect` but `Aff` because the former doesn't allow one to do asynchronous things.

Shouldn't `Application` be redefined to use `Aff` rather then `Effect`?

**Comment by Woody88** [comment]
Yes, I have been thinking a lot about it. The thing is that most of our libs (in ps ecosystem) are still event-based, therefore, most of the APIs are using the `Effect` monad. One quick example would `node-stream` or `node-http`. 

I am not sure if `Aff` does cause any issue in node, it seems to be heavily tested in the frontend. It's also easy to just `runAff` in the `Effect` monad.

However, you are right that if `wai` is to represent only what an `Application` is when talking about client-webserver communication it should be an `Aff`. Would it give us more benefit if we were to use the `MonadAff` constraint instead? 

I'd like to hear your opinion.

**Comment by JordanMartinez** [comment]
> Yes, I have been thinking a lot about it. The thing is that most of our libs (in ps ecosystem) are still event-based, therefore, most of the APIs are using the Effect monad. One quick example would node-stream or node-http.

Wrapping those into `Aff` is relatively straight forward though...

> I am not sure if Aff does cause any issue in node, it seems to be heavily tested in the frontend. It's also easy to just runAff in the Effect monad.

`Aff` doesn't have anything node/browser related AFAIK since it's just JavaScript. So, I don't understand what your concern is here. Is it perhaps that Node.JS might implement things differently than Chrome/Firefox/Edge?

> However, you are right that if wai is to represent only what an Application is when talking about client-webserver communication it should be an Aff. Would it give us more benefit if we were to use the MonadAff constraint instead?

Sorry, I meant we should use `MonadAff`, not `Aff` directly as it would then allow us to use a custom monad more easily.

That being said, here are some other thoughts.

I originally thought we might have to write this:
```purescript
-- `launchAff_` appears in each handler...
app req f = case parse routingDuplexParser req.rawPathInfo of
  Home -> launchAff_ do
    -- ...
  Page1-> launchAff_ do
    -- ...
  Page2-> launchAff_ do
    -- ...
  Page3-> launchAff_ do
    -- ...
```
but then realized that that wasn't necessary if we put it in front of the entire case block:
```purescript
app req f = launchAff_ case parse routingDuplexParser req.rawPathInfo of
  Home -> do
    -- ...
  Page1-> do
    -- ...
  Page2-> do
    -- ...
  Page3-> do
    -- ...
```

I think the other question was how to use custom monad. However, I think that also could be addressed:
```purescript
newtype AppM a = AppM (ReaderT Env Aff a)

app :: Application
app req f = launchAff_ $ runAppM env (app' req f)

app' :: forall m. MonadAff m => _ -> _ -> m Unit
app' req f = case parse routingDuplexParser req.rawPathInfo of
  Home -> do
    -- ...
  Page1-> do
    -- ...
  Page2-> do
    -- ...
  Page3-> do
    -- ...
```

Then I thought that the `Middleware` type might have an issue. For example, if John writes a library that runs in `Effect` and I have my custom monad, how would I use his middleware in my application? I'm asking this question partly because it's been a while since I've done server stuff. So, I can't recall how this works exactly.
```purescript
johnMiddleware :: Middleware
johnMiddleware = -- ...

app :: Application
app req f = launchAff_ $ runAppM env (app' req f)

app' :: forall m. MonadAff m => _ -> _ -> m Unit
app' req f = case parse routingDuplexParser req.rawPathInfo of
  Home -> do
    -- ...
  Page1-> do
    -- ...
  Page2-> do
    -- ...
  Page3-> do
    -- ...
```

Regardless, there is one reason why not using `MonadAff`/`Aff` is a good idea. `Aff` is a pretty crazy piece of JavaScript, so I don't expect other backends to implement it easily.

**Comment by robertdp** [comment]
`IO` in Haskell represents asynchronous effects, because its default execution model is async. JavaScripts default is synchronous, and that maps to `Effect`. `Aff` is really nice to have, especially in the browser, but I would be cautious about requiring its use everywhere especially when not necessarily needed.

Using `Effect` puts a little more constraint and a couple more requirements on the users, but if they understand the sync/async and sequential/parallel effects of transitioning from `Effect -> Aff` and `Aff -> Effect` then it’s easier to understand those boundaries when they are encountered in the library. And it makes those boundaries more explicit:

- `Effect -> Aff` will sequence a sync effect to run after an async effect
- `Aff -> Effect` will run an async effect in “parallel” (not really) but continue immediately (or almost) with the sequence of sync logic.

If everything was written using `Aff` then running logic in parallel would be achieved using `forkAff`, but that certain user-supplied logic will run in parallel is implicit and would require documentation. If using `Effect` then the async parallelism is opt-in and obvious in the users code.

For example: if an event callback uses `Effect`, then it’s clear that it will either run some sync effect immediately, or the user can use `launchAff_` to run an async effect in parallel. But if the callback uses `Aff` does that logic get used sequentially somewhere, blocking the rest of the execution? Or does it run in parallel, and if so how are errors handled and how high up the `Aff` stack can they propagate?

**Update 2020/06/08:** This post is not particularly relevant here, sorry. It came from a discussion about how to design and expose event handlers in a more generic sense, and I was asked to share it here. It doesn't apply to something more specific like the core application type. Feel free to ignore.

**Comment by hdgarrood** [comment]
> The thing is that most of our libs (in ps ecosystem) are still event-based, therefore, most of the APIs are using the Effect monad. One quick example would node-stream or node-http.

I wouldn’t read too much into the purescript-node libraries using Effect; it’s mostly this way because we want these bindings to be as low level and unopinionated as possible to allow other libraries to build on top of them.

Aff works in all JS environments by the way, that’s not a concern. Alternative backends potentially are a concern though. I’m not sure what they tend to do with respect to Effect vs Aff.

**Comment by Woody88** [comment]
This issue is now solved.
